plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1359 @@
+        number of ionic species for the corresponding element as the
+        value.
+        """
+        return self._nstates
+
+    @property
+    def elements(self) -> List[str]:
+        """The elements modeled by this simulation."""
+        return self._elements
+
+    @property
+    def abundances(self) -> Dict[str, float]:
+        """
+        The relative elemental abundances of the elements modeled in
+        this simulation.
+
+        The keys are the atomic symbols and the values are a `float`
+        representing that element's elemental abundance.
+        """
+        return self._abundances
+
+    @property
+    def ionic_fractions(self) -> Dict[str, np.ndarray]:
+        """
+        Return the ionic fractions over the course of the simulation.
+
+        The keys of this dictionary are atomic symbols.  The values are
+        2D arrays where the first index refers to the time step and the
+        second index refers to the integer charge.
+        """
+        return self._ionic_fractions
+
+    @property
+    def number_densities(self) -> Dict[str, u.Quantity]:
+        """
+        Return the number densities over the course of the simulation.
+
+        The keys of ``number_densities`` are atomic symbols.  The values
+        are 2D arrays with units of number density where the first index
+        refers to the time step and the second index is the integer
+        charge.
+
+        """
+        return self._number_densities
+
+    @property
+    def n_elem(self) -> Dict[str, u.Quantity]:
+        """
+        The number densities of each element over the course of the
+        simulation.
+
+        The keys of ``n_elem`` are atomic symbols.  The values are 1D
+        arrays with units of number density where the index refers to
+        the time step.
+
+        """
+        return self._n_elem
+
+    @property
+    def n_e(self) -> u.Quantity:
+        """
+        The electron number density over the course of the simulation in
+        units of number density.
+
+        The index of this array corresponds to the time step.
+        """
+        return self._n_e
+
+    @property
+    def T_e(self) -> u.Quantity:
+        """
+        The electron temperature over the course of the simulation in
+        kelvin.
+
+        The index of this array corresponds to the time step.
+        """
+        return self._T_e
+
+    @property
+    def time(self) -> u.Quantity:
+        """
+        The time for each time step over the course of the simulation
+        in units of seconds.
+        """
+        return self._time
+
+
+class NEI:
+    r"""
+    Perform and analyze a non-equilibrium ionization simulation.
+
+    Parameters
+    ----------
+    inputs
+
+    T_e: astropy.units.Quantity or callable
+        The electron temperature, which may be a constant, an array of
+        temperatures corresponding to the times in `time_input`, or a
+        function that yields the temperature as a function of time.
```

Comment:
Okay, that's pretty cool!

Reply:
This was my favorite part of this whole `NEI` class.  With all that said, that reminds me that it will be necessary to implement a different formulation...

- [ ] Create issue on allowing heat to be added to the parcel of plasma over the course of the simulation
- [ ] Create issue on enabling cooling due to adiabatic expansion
- [ ] Create issue on enabling radiative cooling based on the actual ionization states (this one is going to be a tough one)